src/Cli/dotnet/Commands/Test/MTP/MSBuildUtility.cs

```diff
@@ -18,28 +20,39 @@ internal static class MSBuildUtility
 {
     private const string dotnetTestVerb = "dotnet-test";
 
+    // https://github.com/dotnet/msbuild/pull/7992 :/
+    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "ProjectShouldBuild")]
+    static extern bool ProjectShouldBuild(SolutionFile solutionFile, string projectFile);
+
     public static (IEnumerable<ParallelizableTestModuleGroupWithSequentialInnerModules> Projects, bool IsBuiltOrRestored) GetProjectsFromSolution(string solutionFilePath, BuildOptions buildOptions)
     {
-        SolutionModel solutionModel = SlnFileFactory.CreateFromFileOrDirectory(solutionFilePath, includeSolutionFilterFiles: true, includeSolutionXmlFiles: true);
-
         bool isBuiltOrRestored = BuildOrRestoreProjectOrSolution(solutionFilePath, buildOptions);
 
         if (!isBuiltOrRestored)
         {
             return (Array.Empty<ParallelizableTestModuleGroupWithSequentialInnerModules>(), isBuiltOrRestored);
         }
 
-        string rootDirectory = solutionFilePath.HasExtension(".slnf") ?
-                Path.GetDirectoryName(solutionModel.Description)! :
-                SolutionAndProjectUtility.GetRootDirectory(solutionFilePath);
+        var solutionFile = SolutionFile.Parse(solutionFilePath);
```

Comment:
We would rather you not depend on `SolutionFile`. It's a private implementation detail we make minimal guarantees about. But this is not 100% a deal-breaker for this PR.